### PR TITLE
refactor: Moves simplemob atmos and temperature requirements to DCS elements.

### PR DESCRIPTION
**Pull request by warriorstar-orion**
## What Does This PR Do
This PR moves code for atmospherics and temperature changes for simplemobs to elements. As a refresher, elements are like components except every element is a singleton that the objects it's attached to shares. Bespoke elements allow for variance in initial arguments, making singletons for every unique set of arguments.

## Why It's Good For The Game
This is part of the piecemeal port of TG's basic mobs. It decouples the logic of homeostasis from the basic mob implementation itself, making the mob implementation smaller. In our case, since we don't have basic mobs yet, I've used it to backwards refactor simplemobs as well. This allows us to make sure the code works before the rest of the port.

## Images of changes

<details><summary>Examples of the same element being used by multiple mobs for each element type</summary>

</details>

## Testing
In progress.

- [X] Verified in VV that mobs are being grouped by element per atmos/breathing requirements
- [X] Tested spiders like Araneus took atmos damage from being in a pure plasma environment
- [X] Tested corgis die in space
- [X] Tested space bats don't die in space
- [x] Test slimes and cold weakness / fire affinity

## Changelog
NPFC

### CODE REVIEW

**Comment by BiancaWilkson** [comment]
Slimes and their cold weakness and fire affinity should probably be tested. I can see that possibly causing some weirdness.

**Comment by warriorstar-orion** [comment]
> Slimes and their cold weakness and fire affinity should probably be tested. I can see that possibly causing some weirdness.

Looks good:

2024_06_01__07_23_42__Paradise Station 13
2024_06_01__07_26_06__Paradise Station 13

File: code/datums/elements/atmos_requirements.dm [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	atmos_requirements = atmos_requirements_
+	unsuitable_atmos_damage = unsuitable_atmos_damage_
+	RegisterSignal(target, COMSIG_SIMPLEANIMAL_HANDLE_ENVIRONMENT, PROC_REF(handle_environment))
+
+/datum/element/atmos_requirements/Detach(datum/target)
+	UnregisterSignal(target, COMSIG_SIMPLEANIMAL_HANDLE_ENVIRONMENT)
+	return ..()
+
+/datum/element/atmos_requirements/proc/handle_environment(mob/living/simple_animal/target)
+	SIGNAL_HANDLER
```
Comment (lewcc) [inline code/datums/elements/atmos_requirements.dm:30]:
```suggestion
	SIGNAL_HANDLER  // COMSIG_SIMPLEANIMAL_HANDLE_ENVIRONMENT
```

File: code/datums/elements/body_temperature.dm [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+/**
+ * Bespoke element that deals damage to the attached mob when the ambient temperature is unsuitable.
+ */
+/datum/element/body_temperature
+	element_flags = ELEMENT_BESPOKE
+	id_arg_index = 2
+
+	/// Min body temp
+	var/min_body_temp = NONCARBON_DEFAULT_MIN_BODY_TEMP
+	///Max body temp
```
Comment (lewcc) [inline code/datums/elements/body_temperature.dm:15]:
```suggestion
	/// Max body temp
```

File: code/datums/elements/body_temperature.dm [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+		heat_damage_per_tick = heat_damage_per_tick_
+
+	RegisterSignal(target, COMSIG_SIMPLEANIMAL_HANDLE_ENVIRONMENT, PROC_REF(handle_temperature))
+
+/datum/element/body_temperature/Detach(datum/target)
+	UnregisterSignal(target, COMSIG_SIMPLEANIMAL_HANDLE_ENVIRONMENT)
+	return ..()
+
+/datum/element/body_temperature/proc/handle_temperature(mob/living/simple_animal/target)
+	SIGNAL_HANDLER
```
Comment (lewcc) [inline code/datums/elements/body_temperature.dm:48]:
```suggestion
	SIGNAL_HANDLER  // COMSIG_SIMPLEANIMAL_HANDLE_ENVIRONMENT
```

**Review by lewcc — COMMENTED** [review]
Little nits, code generally looks good.

**Review by Sirryan2002 — APPROVED** [review]
just add the comments lewcc wanted

**Comment by warriorstar-orion** [comment]
> just add the comments lewcc wanted

Ahh this conflicts with a bit of the MILLA changes. I'll look into it.

**Review by lewcc — APPROVED** [review]
LGTM, pending TM

**Comment by SteelSlayer** [comment]
This was TM'd for a round and this was runtime was happening quite a lot. Pulling the TM for now.

```
[2024-07-11T04:25:57] Runtime in code/controllers/subsystem/processing/SSdcs.dm:38: list index out of bounds
   proc name: GetIdFromArguments (/datum/controller/subsystem/processing/dcs/proc/GetIdFromArguments)
   src: Datum Component System (/datum/controller/subsystem/processing/dcs)
   call stack:
   Datum Component System (/datum/controller/subsystem/processing/dcs): GetIdFromArguments
   Datum Component System (/datum/controller/subsystem/processing/dcs): GetElement
   Pirate Gunner (/mob/living/simple_animal/hostile/pirate/ranged): _RemoveElement
   Pirate Gunner (/mob/living/simple_animal/hostile/pirate/ranged): remove_atmos_requirements
   Pirate Gunner (/mob/living/simple_animal/hostile/pirate/ranged): Destroy
   Pirate Gunner (/mob/living/simple_animal/hostile/pirate/ranged): Destroy
   qdel(Pirate Gunner (/mob/living/simple_animal/hostile/pirate/ranged), 0)
   Pirate Gunner (/mob/living/simple_animal/hostile/pirate/ranged): death
   Pirate Gunner (/mob/living/simple_animal/hostile/pirate/ranged): death
   Pirate Gunner (/mob/living/simple_animal/hostile/pirate/ranged): update_stat
   ...
   Pirate Gunner (/mob/living/simple_animal/hostile/pirate/ranged): adjustBruteLoss
   Pirate Gunner (/mob/living/simple_animal/hostile/pirate/ranged): apply_damage
   Pirate Gunner (/mob/living/simple_animal/hostile/pirate/ranged): bullet_act
   Pirate Gunner (/mob/living/simple_animal/hostile/pirate/ranged): bullet_act
   the kinetic force (/obj/item/projectile/kinetic): Bump
   the kinetic force (/obj/item/projectile/kinetic): Move
   the kinetic force (/obj/item/projectile/kinetic): process
   Projectiles (/datum/controller/subsystem/processing/projectiles): ignite
   Master (/datum/controller/master): Loop
   Master (/datum/controller/master): StartProcessing
```
```
[2024-07-11T05:00:06] Runtime in code/controllers/subsystem/processing/SSdcs.dm:38: list index out of bounds
   proc name: GetIdFromArguments (/datum/controller/subsystem/processing/dcs/proc/GetIdFromArguments)
   usr: Lawliet (CKEY) (/mob/living/carbon/human)
   usr.loc: The floor (167,87,2) (/turf/simulated/floor/plasteel)
   src: Datum Component System (/datum/controller/subsystem/processing/dcs)
   call stack:
   Datum Component System (/datum/controller/subsystem/processing/dcs): GetIdFromArguments
   Datum Component System (/datum/controller/subsystem/processing/dcs): GetElement
   the dark purple baby slime (35... (/mob/living/simple_animal/slime): _RemoveElement
   the dark purple baby slime (35... (/mob/living/simple_animal/slime): remove_atmos_requirements
   the dark purple baby slime (35... (/mob/living/simple_animal/slime): Destroy
   the dark purple baby slime (35... (/mob/living/simple_animal/slime): Destroy
   qdel(the dark purple baby slime (35... (/mob/living/simple_animal/slime), 0)
   /datum/food_processor_process/... (/datum/food_processor_process/mob/slime): process_food
   /datum/food_processor_process/... (/datum/food_processor_process/mob/slime): process_food
   /datum/food_processor_process/... (/datum/food_processor_process/mob/slime): process_food
   Slime Processor (/obj/machinery/processor): attack_hand
   Lawliet (/mob/living/carbon/human): UnarmedAttack
   Lawliet (/mob/living/carbon/human): ClickOn
   Slime Processor (/obj/machinery/processor): MouseDrop
```

**Comment by S34NW** [comment]
This PR has been marked as stale due to lack of activity, please continue working on the PR and/or address any comments within 7 days or we will have to close it.

**Comment by warriorstar-orion** [comment]
This was broken by https://github.com/ParadiseSS13/Paradise/pull/25972 which snuck in before it, I've been waiting for @chuga-git to pin down the issue. I'll do a merge master in the meantime.

**Comment by S34NW** [comment]
https://github.com/ParadiseSS13/Paradise/pull/26387 should fix runtimes here

**Comment by Danchi299** [comment]
class components?
in BYOND programming?
that is surely interesting

File: code/modules/mob/living/simple_animal/simple_animal.dm [hunk 4]
```diff
@@ -181,8 +184,23 @@
...
+	RemoveElement(/datum/element/body_temperature)
+
 	return ..()
 
+/mob/living/simple_animal/proc/apply_atmos_requirements()
+	if(unsuitable_atmos_damage == 0)
+		return
+
+	// String assoc list returns a cached list, so this is like a static list to
+	// pass into the element below.
```
Comment (Burzah) [inline code/modules/mob/living/simple_animal/simple_animal.dm:197]:
```suggestion
	/*
	*  String associated list returns a cached list. 
	*  This is like a static list to pass into the element below.
	*/
```

**Review by Burzah — APPROVED** [review]
This has been on TM for awhile now after it was adjusted, with no reported issues. Looks good to me.

**Pull request merged**